content/blog/provisioned-concurrency-insurance-against-cold-starts-in-aws-lambda/index.md

```diff
@@ -0,0 +1,154 @@
+<!--more-->
+
+## Concurrency Model of AWS Lambda
+
+Despite being serverless, AWS Lambda uses lightweight containers to process incoming requests. Every container, or worker, can process only a single request at any given time.
+
+{{< figure src="./executions.png" caption="Overlapping executions land on separate workers" >}}
+
+Therefore, the number of concurrent requests define the number of required workers that a specific AWS Lambda needs to serve at any given moment.
+
+## Cold Starts
+
+How does AWS know how many workers it needs to run for a given function? Well, it doesn't know in advance. AWS allocates new workers on-demand as the Lambda gets invoked.
+Whenever Lambda receives a request but it has no idle workers, the control plane assigns a new generic worker to it. The worker then has to download the custom code or binaries of your Lambda and load them into memory before it can do service the request. This process takes time, which significantly increases response latency.
+
+The issue of sporadically slow responses caused by the need to increase the pool of workers is known as **Cold Start**. Cold starts are consistently the top concern about the applicability of serverless tech to latency-sensitive workloads. There are numerous articles about the problem, including many articles I have written in the [Cold Starts](https://mikhail.io/serverless/coldstarts/) section on my website
+
+## Warming
+
+Cold starts don't occur for the majority of requests because AWS Lambda reuses workers between subsequent invocations. However, if a particular worker is idle for too long (usually, several minutes), AWS may decide to recycle and return it to the generic pool.
+
+A trick known as **Lambda Warmers** uses kept-alive workers to reduce the frequency of cold starts. The idea is to have a CloudWatch timer that fires every few minutes and sends `N` parallel requests to the target Lambda. If all those requests land at the same time, AWS has to provision at least `N` workers to process them. The actual operation doesn't have to do any useful work, but it resets the recycling timer back to zero.
+
+There are a few drawbacks with this approach:
+
+* If a valid request comes at the same time as warming requests, it might hit a cold start.
+* Extra logic is needed in the Lambda code to detect warming requests and short-circuit instead of doing useful work.
+* CloudWatch rules require additional setup and management.
+* The result is still best-effort: Lambda still recycles workers from time to time.
+
+You can find more details about how Pulumi can help with some of these issues in [AWS Lambda Warmer as Pulumi Component](https://mikhail.io/2018/08/aws-lambda-warmer-as-pulumi-component/).
+
+Here is an example of configuring the provisioned concurrency with Pulumi in TypeScript:
```

Comment:
This is a bit of an abrupt transition - shouldn't there be a section header with an introduction to what Provisioned Concurrency is, how it works, etc? and then the code example?

Reply:
Wow, definitely not intended. A title and a paragraph from my original doc are gone. Restoring!

Reply:
I don't see the transition in the google doc across versions.